sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -344,10 +344,11 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
       new ResolveHints.RemoveAllHints),
     Batch("Nondeterministic", Once,
       PullOutNondeterministic),
+    Batch("ScalaUDF", Once,
+      HandleNullInputsForUDF),
```

Comment:
This rule adds null handling for ScalaUDF, so even if the `ScalaUDF`'s `nullable` is false, the final expression (if-else) can also return null, so we need to run `UpdateAttributeNullability` after it.

Reply:
But we also need to run it before? Or how could we correctly handle null inputs to udf if we do not have the correct nullability for the inputs?